### PR TITLE
fix($animate): avoid memory leak with `$animate.enabled(element, enabled)`

### LINKED ISSUE #16637: ngAnimate: memory leak in disabledElementsLookup Map
All begins with this line of the code:

https://github.com/angular/angular.js/blob/494277e3d3668699133fc796e5bf632258e34802/src/ngAnimate/animateQueue.js#L306

**Reproducable:** always
**Browsers:** Any
**Operating system:** Any
**Versions**: All that use the hash map

**Steps to reproduce:**

Available in codepen: https://codepen.io/anon/pen/yqVwwE

click on "leaker" **n** times to get **n** detached nodes as seen in the screenshot

screen shot 2018-07-20 at 12 37 41

**Solution**: listen for element scope destroy event, or ``element.on('remove')`` and delete the node from ``disabledElementsLookup`` Map ?

**Note**: If this is "working as intended" at least provide public api to ``disabledElementsLookup`` so we can cleanup after directive exits

### PR DESCRIPTION
**Pull request by gkalpak**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix (memory leak).

**What is the current behavior? (You can also link to an open issue here)**
When disabling/enabling animations on a specific element (via `$animate.enabled(element, enabled)`), the element is added in a map to track its state.

Currently, the element is never removed from the map, causing AngularJS to hold on to the element even after it is removed from the DOM, thus preventing it from being garbage collected.

See #16637.

**What is the new behavior (if this is a feature change)?**
The element is removed from the map on `$destroy`, making it eligible for garbage collection (if there are no references to it).

**Does this PR introduce a breaking change?**
<sub>Maybe... (bar probably not really)</sub>

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

**Other information**:
Fixes #16637.

### CODE REVIEW

**Comment by gkalpak** [comment]
It crossed my ind that this migth be a breaking change, because previously removing and re-inserting an element let it retain its "animations enabled" status. 

But generally a lot of things will break if you remove an element (via `.remove()`), since all data and event listeners are removed as well (from a jqLite/jQuery perspective, not native ones).

The proper method if you want to temporarily remove an element from the DOM is `.detach()` (which retains data and listeners).

##
An alternative would be to attach the "animations enabled" status on the element itself (directly, not via `.data()` for the same reasons).

##
I lean towards not considering this a breaking change, though 😅 WDYT?

**Comment by Narretz** [comment]
I don't think it's a breaking change. It was an interesting side-effect but if somebody relied on this behavior I don't think it's many people.

**Review by Narretz — APPROVED** [review]

**Review by mgol — APPROVED** [review]

**Comment by gkalpak** [comment]
There was a bug (with jQuery) :grin: Pushed a fixup commit.
I'll merge as soon as CI is green(-ish).
